pdns/recursordist/rec-taskqueue.cc

```diff
@@ -22,50 +22,129 @@
 #include "rec-taskqueue.hh"
 #include "taskqueue.hh"
 #include "lock.hh"
-#include "logger.hh"
+#include "logging.hh"
 #include "stat_t.hh"
 #include "syncres.hh"
 
-static LockGuarded<pdns::TaskQueue> s_taskQueue;
-static pdns::stat_t s_almost_expired_tasks_pushed;
-static pdns::stat_t s_almost_expired_tasks_run;
-static pdns::stat_t s_almost_expired_tasks_exceptions;
+// For rate lmiting, we maintain a set of tasks recently submitted.
+class TimedSet
+{
+public:
+  TimedSet(time_t t) :
+    d_expiry_seconds(t)
+  {
+  }
+  bool insert(time_t now, const pdns::ResolveTask& task)
+  {
+    time_t ttd = now + d_expiry_seconds;
+    bool inserted = d_set.emplace(task, ttd).second;
+    if (!inserted) {
+      // Instead of a periodic clean, we always do it on a hit
+      // the operation should be cheap as we just walk the ordered time_t index
+      // There is a slim chance if we never hit a rate limiting case we'll never clean... oh well
```

Comment:
I'm a bit worried about the odds of that happening in some setups, because without collision the set will grow endlessly..